websites/D/Duolingo/presence.ts

```diff
@@ -1,92 +1,232 @@
-lang.set("nb", "Norwegian");
-lang.set("vi", "Vietnamese");
-lang.set("ar", "Arabic");
-lang.set("hw", "Hawaiian");
-lang.set("da", "Danish");
-lang.set("kl", "Klingon");
-lang.set("ro", "Romanian");
-lang.set("cs", "Czech");
-lang.set("sw", "Swahili");
-lang.set("cy", "Walsh");
-lang.set("id", "Indonesian");
-lang.set("hu", "Hungarian");
-lang.set("uk", "Ukrainian");
-lang.set("eo", "Esperanto");
-lang.set("nv", "Navajo");
-lang.set("en", "English");
+  language = { imageKey: "icon", name: "nothing (yet!)" };
 
-presence.on("UpdateData", async () => {
-  const path1 = document.location.pathname;
+function updateLanguage() {
+  const state = JSON.parse(window.localStorage.getItem("duo.state")),
+    courseId = state?.user?.currentCourseId;
+  if (!courseId) return;
+
+  const course = state.courses?.[courseId];
+  if (!course) return;
 
-  if (!path1.split("/")[2]) {
-    if (document.location.pathname.startsWith("/learn")) {
-      pageData.details = "Choosing level to learn..";
-      presence.setActivity(pageData);
-    } else if (document.location.pathname.startsWith("/shop")) {
-      pageData.details = "Browsing shop..";
-      presence.setActivity(pageData);
-    } else if (document.location.pathname.includes("/dictionary")) {
-      pageData.details = "Using dictionary..";
-      pageData.state = `Language: ${document.location.pathname.split("/")[2]}`;
-      presence.setActivity(pageData);
-    } else if (document.location.pathname.includes("/profile")) {
-      pageData.details = "Browsing profile..";
-      pageData.state = `Browsing: ${document.location.pathname.split("/")[2]}`;
-      presence.setActivity(pageData);
-    } else if (document.location.pathname.includes("/words")) {
-      presenceData.details = "Checking words...";
-      presenceData.largeImageKey = "logo";
-      presence.setActivity(presenceData);
-    } else if (
-      document.location.pathname === "/" ||
-      !document.location.pathname
-    )
-      presence.setActivity(pageData);
-  } else {
-    if (
-      path1.length > 1 &&
-      path1.split("/")[2] !== null &&
-      path1.split("/")[2].length === 2
-    ) {
-      let language: string;
-      for (const value of lang.keys()) {
-        if (path1.split("/")[2] === value) {
-          language = lang.get(value);
-          break;
-        }
-      }
-      presenceData.details = "Taking a lesson";
-      presenceData.state = `Language: ${language}`;
-      presenceData.largeImageKey = "logo";
-      presence.setActivity(presenceData);
+  const languageId = course.learningLanguage;
+  if (FLAG_ICONS.includes(languageId)) language.imageKey = `flag_${languageId}`;
+  else if (languageId) language.imageKey = "flag_unknown";
+  else language.imageKey = "icon";
+  language.name = course.title ?? "nothing (yet!)";
+}
+updateLanguage();
+setInterval(updateLanguage, 1000);
+
+function checkBasicPages(path: string[]) {
+  if (path.length === 0) {
+    const url = new URL(document.location.href);
```

Comment:
bad practice to only use a variable once

Reply:
Okay I feel like that's highly debatable. That looks WAY more readable than:

```js
if (new URL(document.location.href).searchParams.get("isLoggingIn") === "true") {
```

And plus it keeps that line shorter. If you still insist on that change though, I'm fine with that.

Reply:
We have made a ESLint plugin to prevent on-time-vars which will be implemented in #4893

Reply:
ughhh okay. I feel like that's a really bad decision and that it would be better to prioritize code readability over whatever else you're prioritizing, but that's not up to me so I'll change it.

Reply:
> Okay I feel like that's highly debatable. That looks WAY more readable than:
> 
> ```js
> if (new URL(document.location.href).searchParams.get("isLoggingIn") === "true") {
> ```
> 
> And plus it keeps that line shorter. If you still insist on that change though, I'm fine with that.

in some cases like in ifs it might look uglier, but there are a LOT of "useless" variables in the presences. You can always add a comment above to if to clarify what it does, personally I think the if in your example is pretty readable, checking whether the search param has isLoggingIn to true. But since you are not the only contributor to this presence it might be nice to add a comment if you wish to do so.

Reply:
> in some cases like in ifs it might look uglier, but there are a LOT of "useless" variables in the presences.

No yeah I don't doubt that, but I feel like there are still a lot of cases where it's appropriate to create one-time variables to make the code more self-explanatory or to make it look nicer.

> You can always add a comment above to if to clarify what it does, personally I think the if in your example is pretty readable, checking whether the search param has isLoggingIn to true.

In that example I think the code is still self-explanatory, I just think it's harder to look at than before because everything is stuffed into a single line.

I just don't think it's a good idea to disallow one-time variables because there are many instances where it does clarify what code does and make it look much nicer. Yeah you could just use comments, but why do that when you can just create a separate variable for it which also has the added bonus of looking nicer?